---
title: "DOQTL Workshop: Short Course on Systems Genetics 2015"
author: "Daniel M. Gatti"
date: "10/1/2015"
output: 
  html_document: 
    keep_md: yes
    toc: yes
---

## Introduction 

We will be mapping phenotypes from unpublished data using [Diversity Outbred (DO)  Mice](http://jaxmice.jax.org/strain/009376.html). 192 DO mice (98 females & 94 males) were
placed on either a chow diet or a high fat diet. They were aged to approximately 22 weeks
and a variety of physiological phenotypes were measured. At sacrifice, we collected liver tissue 
and measured gene expression using RNASeq and protein levels using mass spectrometry.

Sex | chow | HFD
----|------|----
 F  |  50  |  48

In this tutorial, we will map gene expression traits. We will perform linkage mapping using the DO haplotype reconstructions and then perform association mapping using imputed SNPs from the DO founders. You will then map an expression trait that you choose.

![DO figure](./DO.circle.figure.png)

## Libraries and Data Import

First, we will load the R libraries that we need to perform the analysis. These include [DOQTL](http://bioconductor.org/packages/release/bioc/html/DOQTL.html), a package designed to map traits in DO mice.

```{r, load_doqtl,results='hide',message=FALSE}
library(DOQTL)
```

We will be working with data that is stored locally. For the first exercise, we will be mapping a gene expression trait. In order to map, we need two pieces of information on each mouse:

1. Phenotype measurements on a set of DO mice
2. Genotypes across the genome on the same set of DO mice

All of the data for this tutorial is stored in a compressed R binary file that ends with "Rdata". You load an Rdata file in using the *load()* command.

Next, we need to load in the haplotype probabilities. These are in large Rdata files with other variables that will be used later.

```{r load_data,warning=FALSE}
load("/data/Rdata/DO192_DataforSysGenCourse.Rdata")
```

These commands loaded a number of R objects into the environment. Look at the "Environment" tab in the upper left panel of your RStudio window to see them.

We are going to remove a few of the objects that we will not use today.

```{r remove_extra_data,warning=FALSE}
rm(annotations.protein.192, covariates.protein.192, expr.protein.192)
```

First, let's look at the expression data in "expr.rna.192".  It is a large matrix with `r nrow(expr.rna.192)` rows and `r ncol(expr.rna.192)` columns. You are required to have the sample IDs in the rownames of your phenotype type data. Let's look at a small part of the file.

```{r look_expr,warnings=FALSE}
expr.rna.192[1:5,1:6]
```

The rows contain sample IDs (F326, F327, F329, etc.) and the columns contain Ensembl Gene IDs. In order to map one gene, we would use one coulmn of data.

Next, let's look at the distribution of one of the genes.

```{r hist_expr,warning=FALSE}
hist(expr.rna.192[,1], breaks = 20, col = "grey", main = colnames(expr.rna.192)[1])
```

Amazing! It lookes normally distributed. In fact, we have transformed each trait using an inverse normal (or rankZ) transformation. DOQTL contains a function called "rankZ()" that has been applied to each gene. If we leave the exrpession data untransformed, we may have outliers that would adversely affect the QTL mapping. The rankZ transformation provides a compromise between no transformation and simply ranking the data. Value at the extremes of the distriubtion will still have more leverage than values near the mean, but not enough to disrupt the mapping. Also, when all traits have the same distribution we can run permutations on only one gene to assess significance thresholds. 

Next, let's look at the haplotype probabilities in "probs.192". "probs.192"" is a three dimensional numeric array. The `r nrow(probs.192)` samples are in rows, the `r ncol(probs.192)` DO founders are in columns and the `r dim(probs.192)[3]` markers are in slices. Again, as with the expression data, you are required to have sample IDs in rownames.

![Haplotype Probs](./haplotype_probs.png)

Let's look at one slice of the haplotype probabilities, at one SNP for the first 20 samples.

```{r haplo_image1,warning=FALSE,fig.height=8}
image(1:ncol(probs.192), 1:20, t(probs.192[20:1,,1]), axes = F, ann = F,
      breaks = c(-0.25, 0.25, 0.75, 1.25), col = c("white", "grey50", "black"))
box()
abline(v = 0:9+0.5, col = "grey80")
abline(h = 0:20+0.5, col = "grey80")
mtext(side = 3, line = 0.5, at = 1:8, text = LETTERS[1:8], cex = 1.5)
mtext(side = 2, line = 0.5, at = 20:1, text = rownames(probs.192)[1:20], las = 1)
col = rep(rgb(0,0,0), length(probs.192[1:20,,1]))
col[which(t(probs.192[20:1,,1]) > 0.9)] = rgb(1,1,1)
text(x = rep(1:8, 20), y = rep(1:20, each = 8), labels = round(t(probs.192[20:1,,1]), digits = 1), col = col)
```

The eight founders are listed in columns and 20 samples are listed in each row.  The values in the cells are either 0, 0.5 or 1, indicating that a sample contains 0, 1 or 2 copies of each founder allele. For example, sample F329 in the first row carries one copy of the "A" allele and one copy of the "F" allele because both the "A" and "F" columns contain 0.5.  In contrast, sample F338 carries 2 copies of the "D" allele because there is a 1 in the "D" column.

Let's look at a different slice of the haplotype probabilities, at the first 100 SNPs for **one** sample.

```{r haplo_image2,warning=FALSE,fig.height=8}
image(1:ncol(probs.192), 1:500, probs.192[1,,500:1], axes = F, ann = F,
      breaks = c(-0.25, 0.25, 0.75, 1.25), col = c("white", "grey50", "black"))
box()
abline(v = 0:9+0.5, col = "grey80")
mtext(side = 3, line = 0.5, at = 1:8, text = LETTERS[1:8], cex = 1.5)
mtext(side = 2, line = 0.5, at = 0:4*100, text = paste("SNP", 5:1*100), las = 1)
```

In this view, we are looking at one sample across the first 500 markers on chromosome one.  This sample is heterozygous for "AF" from marker 1 to about marker 270, then it is heterozygous for "CF" and then "FG".


We also have a kinship matrix that was produced from the haplotype probabilities. A kinship matrix contains the degree of relatedness between each pair of mice. In this case, we have created a kinship matrix for each chromosome. We calculate the kinship matrix on each chromosome using the markers from all of the other chromosomes, but excluding the markers on the current chromosome. For example, on Chr 1, we would calculate the kinship between mice using markers on Chr 2 through Chr X.  This has been shown to increase power for some QTL. The acronym "LOCO" stands for "Leave One Chromosome Out". Below is example code (that we do not run) showing how to calculate the kinship matrix.

```{r kinship,eval=FALSE}
K = kinship.probs(probs = probs.192, snps = snps.64K, bychr = TRUE)
```

## Mapping a Gene Experssion Trait

We will first create a matrix of covariates to use in the mapping model. DOQTL requires a covariate called "sex" because it is used to map on the X chromosome. We will also use diet as a covariate. There are three important points to remember with covariates.

1. Covariates must be numeric,
2. Covariates must have rownames that contain the same sample IDs as the phenotypes,
3. Covariates must have one column called "sex".

```{r covar,warning=FALSE}
covar = model.matrix(~Sex + Diet, data = covariates.rna.192)[,-1]
colnames(covar)[1] = "sex"
rownames(covar) = rownames(expr.rna.192)
```

Next, we need to decide which gene to map. We will map *Lrtm1* and we need to find the column number in "expr.rna.192" that contains this gene. We will look it up in the "annotations.rna.192" data.frame.

```{r pick_gene,warning=FALSE}
row = which(annotations.rna.192$Gene == "Lrtm1")
ensid = annotations.rna.192$EnsemblID[row]
annotations.rna.192[row,]
```

The last line printed out the annotation data for Lrtm1. What chromosome is Lrtm1 on?  And what Mb position?

We can see that the Ensembl ID for *Lrtm1* is `r annotations.rna.192$EnsemblID[row]`. Next we need to find that column in "expr.rna.192".

```{r pick_gene2,warning=FALSE}
pheno.col = which(colnames(expr.rna.192) == ensid)
colnames(expr.rna.192)[pheno.col]
```

The main function for linkage mapping, in which we regress the phenotype on the haplotype probabilities, is called "scanone()". scanone() uses the sample IDs in the phenotype, covariates, probs and kinship matrices to line up the samples. This means that sample IDs are required in the rownames of all of these objects.

```{r scanone,warning=FALSE}
qtl = scanone(pheno = expr.rna.192, pheno.col = pheno.col, probs = probs.192, K = K.LOCO.192,
             addcovar = covar, snps = snps.64K)
```

This produces a doqtl object that contains the LOD scores and the founder allele effects. Next, we plot the genome scan.

```{r plot_qtl,warning=FALSE}
plot(qtl, main = paste(ensid, "Lrtm1"))
```

There are some peaks that look large, but we need some way to decide which ones are significant. To do this, we will load in previously computed permutations. We ran 1000 permutations and retained the maximum LOD score from each one.

```{r load_perms,warning=FALSE}
load("/data/eQTL/eQTL_perms.Rdata")
thr = quantile(perms, 0.95)
```

This is the distribution of the maximum LOD scores across the permutations. The red line is the 95th percentile of the distribution. LOD scores that are above this threshold will occur by change 5% of the time. We call this an alpha = 0.05 genome-wide threshold.

```{r hist_perms,warning=FALSE}
hist(perms, breaks = 20)
abline(v = thr, col = "red", lwd = 2)
```

Let's add the 0.05 threshold to the QTL plot.

```{r plot_qtl_perms,warning=FALSE}
plot(qtl, main = paste(ensid, "Lrtm1"), sig.thr = thr)
```

We can see that two peaks cross the threshold, one on Chr 7 and one on Chr 14. Let's look more closely at the peak on Chr 14. The mapping model produces an estimate of the effect of each of the eight founder alleles. We can plot these on one chromosome using "coefplot()".


```{r coefplot1,warning=FALSE}
coefplot(qtl, chr = 14, main = paste(ensid, "Lrtm1"))
```

Each of the eight colored lines in the top plot represent the effect of one of the eight founders. The LOD plot is shown on the bottom. At the peak near 30Mb, the allele effects split into two groups. DO mice that carry the PWK/PhJ or WSB/EiJ alleles have lower *Lrtm1* expression than other DO mice. Look to the right near 60 Mb. The founder effects separate from each other.  But is this spilt significant? What would you use to decide if there is another peak at 60 Mb?

*Lrtm1* is located on Chr 14 at 29 Mb. Notice that the peak on Chr 14 is very close to this location. We call this a "local-eQTL" because the QTL peak is co-located with the gene itself.

Next, let's look at the founder effects for the peak on Chr 7.

```{r coefplot2,warning=FALSE}
coefplot(qtl, chr = 7, main = paste(ensid, "Lrtm1"))
```

The peaks here is wider and the pattern of allele effects is less clear.  When a peak is located far from the gene, we call this a "distant eQTL".

Next, we can use the haplotype reconstructions to impute the DO founder SNPs onto DO genomes. We do this by taking SNPs from the [Sanger Mouse Genomes Project](http://www.sanger.ac.uk/resources/mouse/genomes/) and pasting them into the haplotype blocks for each DO mouse.

![DO impute SNPs](./DO.impute.founders.png)

Although we genotyped the DO mice at ~77,000 markers, we now have 40 million SNPs in the DO mice. We can use these to perform association mapping at each SNP. While it is possible to perform genome-wide association mapping, this requires more time than we have in this tutorial. Here will will perform association mapping in the region of the local eQTL for *Lrtm1* on Chr 14. To do this, we will "assoc.map()", which takes a set of arguments similar to "scanone()".

```{r gwas,warning=FALSE}
assoc = assoc.map(pheno = expr.rna.192, pheno.col = pheno.col, probs = probs.192, K = K.LOCO.192[[14]],
                  addcovar = covar, snps = snps.64K, chr = 14, start = 28, end = 32)
tmp = assoc.plot(assoc, thr = 10, show.sdps = TRUE, highlight = "Lrtm1")
```

There are three panels in the association mapping plot. The top panel shows the minor allele frequency for selected SNPs, the middle panel shows the LOD scores for all SNPs, and the bottom panel shows the genes in the interval. We colored SNPs with LOD scores over 10 in red and then plotted the minor allele for each of these SNPs. A LOD of 10 is arbitrary in this case, but it allows us to highlight the most significant SNPs. Which two strains contribute the minor allele for the high LOD SNPs? How do these strains relate to the ones in the founder effects plot above?

## Mapping your favorite gene (that is expressed in the liver)

Next, you will map a gene of your choosing. Pick a gene and see if it is in the data set.


```{r user_exercise,eval=FALSE}
row = which(annotations.rna.192$Gene == "MyFavoriteGene")
ensid = annotations.rna.192$EnsemblID[row]
annotations.rna.192[row,]
pheno.col = which(colnames(expr.rna.192) == ensid)
```

If you got a row containing annotation data, then your gene is in the DO liver expression data.

Try to modify the QTL mapping code from *Lrtm1* to perform the following tasks:

1. Map the gene.
2. Create a QTL plot. 
3. Create a QTL plot with the significance threshold.
4. Create a founder effects plot on the chromosome with the highest LOD score.
5. Perform association mapping near the highest QTL peak .
6. Create an association mapping plot with your favorite gene highlighed (if it is in the QTL interval).

